lib/AST/DeclHead.cpp

```diff
@@ -425,6 +427,195 @@ TemplateParameterList::Parameters(void) const noexcept {
+    return std::nullopt;
+  }
+
+  const clang::Type *type = record->getTypeForDecl();
+  clang::NamedDecl *nd = nullptr;
+  if (type->isIncompleteType(&nd)) {
+    return std::nullopt;
+  }
+
+  auto &layout = record->getASTContext().getASTRecordLayout(record);
+  return layout.getNonVirtualSize().getQuantity();
+}
+
+std::optional<CXXRecordDecl> CXXRecordDecl::PrimaryBase(void) const noexcept {
+  clang::RecordDecl *record = u.RecordDecl->getDefinition();
+  if (!record || record->isInvalidDecl() || !record->isCompleteDefinition()) {
+    return std::nullopt;
+  }
+
+  auto &layout = record->getASTContext().getASTRecordLayout(record);
+  const clang::CXXRecordDecl *base = layout.getPrimaryBase();
+  if (!base) {
+    return std::nullopt;
+  }
+
+  return DeclBuilder::Create<pasta::CXXRecordDecl>(ast, base);
+}
+
+std::optional<bool> CXXRecordDecl::HasOwnVirtualFunctionTablePointer(void) const noexcept {
+  clang::RecordDecl *record = u.RecordDecl->getDefinition();
+  if (!record || record->isInvalidDecl() || !record->isCompleteDefinition()) {
+    return std::nullopt;
+  }
+
+  auto &layout = record->getASTContext().getASTRecordLayout(record);
+  return layout.hasOwnVFPtr();
+}
+
+std::optional<bool> CXXRecordDecl::HasExtendableVirtualFunctionTablePointer(void) const noexcept {
+  clang::RecordDecl *record = u.RecordDecl->getDefinition();
+  if (!record || record->isInvalidDecl() || !record->isCompleteDefinition()) {
+    return std::nullopt;
+  }
+
+  auto &layout = record->getASTContext().getASTRecordLayout(record);
+  return layout.hasExtendableVFPtr();
+}
+
+std::optional<bool> CXXRecordDecl::HasVirtualBaseTablePointer(void) const noexcept {
+  clang::RecordDecl *record = u.RecordDecl->getDefinition();
+  if (!record || record->isInvalidDecl() || !record->isCompleteDefinition()) {
+    return std::nullopt;
+  }
+
+  auto &layout = record->getASTContext().getASTRecordLayout(record);
+  return layout.hasVBPtr();
+}
+
+std::optional<bool> CXXRecordDecl::HasOwnVirtualBaseTablePointer(void) const noexcept {
+  clang::RecordDecl *record = u.RecordDecl->getDefinition();
+  if (!record || record->isInvalidDecl() || !record->isCompleteDefinition()) {
+    return std::nullopt;
+  }
+
+  auto &layout = record->getASTContext().getASTRecordLayout(record);
+  return layout.hasOwnVBPtr();
+}
+
+std::optional<uint64_t> FieldDecl::OffsetInBits(void) const noexcept {
+  const clang::RecordDecl *record = u.FieldDecl->getParent();
+  if (!record) {
+    return std::nullopt;
+  }
+
+  record = record->getDefinition();
+  if (!record || record->isInvalidDecl() || !record->isCompleteDefinition()) {
+    return std::nullopt;
+  }
+
+  const clang::Type *type = record->getTypeForDecl();
+  clang::NamedDecl *nd = nullptr;
+  if (!type || type->isIncompleteType(&nd)) {
+    return std::nullopt;
+  }
+
+  auto &layout = record->getASTContext().getASTRecordLayout(record);
+  return layout.getFieldOffset(u.FieldDecl->getFieldIndex());
```

Comment:
All of ^ contains a lot of straight up C&P or code that can be instead extracted into more composable helpers. I outlined a few, refactor.